thorlcr/activities/indexwrite/thindexwriteslave.cpp

```diff
@@ -212,12 +215,15 @@ class IndexWriteSlaveActivity : public ProcessSlaveActivity, public ILookAheadSt
         if (metadata->getPropBool("_useTrailingHeader", true))
             flags |= USE_TRAILING_HEADER;
         unsigned twFlags = isUrl(partFname) ? TW_Direct : TW_RenameToPrimary;
-        builderIFileIO.setown(createMultipleWrite(this, partDesc, 0, twFlags, compress, NULL, this, &abortSoon));
-        Owned<IFileIOStream> out = createBufferedIOStream(builderIFileIO, 0x100000);
-        if (!needsSeek)
-            out.setown(createNoSeekIOStream(out));
-        maxRecordSizeSeen = 0;
-        builder.setown(createKeyBuilder(out, flags, maxDiskRecordSize, nodeSize, helper->getKeyedSize(), isTlk ? 0 : totalCount, helper, defaultIndexCompression, !isTlk, isTlk));
+        {
+            CriticalBlock b(builderFileCS);
```

Comment:
it may not be that important, but locking scope is quite large here, e.g. could block in createMultipleWrite in i/o etc.

See other comment (and original JIRA description), it would be cleaner if the stats came from the builder rather than from a stashed IFileIO + some from builder, and the mutex to protect gatherActiveStats can then just be around the assignment/clearing of the builder.